In [17]:
def f(x):
    return (x-15)*(x-15)+5

# Дихотомия

In [18]:
print("Метод Дихотомии, исходная функция (x - 15)^2 + 5")
a = 12
b = 20

epsilon = 0.01

func = lambda x: ((x - 15)**2) + 5

count = 0
while (b - a) > (2 * epsilon):
    count += 1
    print(f"\n{count} итерация (a = {a}, b = {b}):\n")
    c = (a + b) / 2
    print(f"c = {c}")
    x1 = c - (epsilon / 2)
    print(f"x1 = {x1}")
    x2 = c + (epsilon / 2)
    print(f"x2 = {x2}")
    f1 = func(x1)
    print(f"f(x1) = {f1}")
    f2 = func(x2)
    print(f"f(x2) = {f2}")
    print(f"l = {b - a}")
    if f1 > f2:
        a = x1
    else:
        b = x2

print(f"\nРезультат: x* = {(a + b) / 2} f* = {func((a + b) / 2)}")

Метод Дихотомии, исходная функция (x - 15)^2 + 5

1 итерация (a = 12, b = 20):

c = 16.0
x1 = 15.995
x2 = 16.005
f(x1) = 5.990024999999998
f(x2) = 6.010024999999998
l = 8

2 итерация (a = 12, b = 16.005):

c = 14.0025
x1 = 13.997499999999999
x2 = 14.0075
f(x1) = 6.005006250000003
f(x2) = 5.9850562499999995
l = 4.004999999999999

3 итерация (a = 13.997499999999999, b = 16.005):

c = 15.001249999999999
x1 = 14.996249999999998
x2 = 15.00625
f(x1) = 5.0000140625
f(x2) = 5.0000390625
l = 2.0075000000000003

4 итерация (a = 13.997499999999999, b = 15.00625):

c = 14.501874999999998
x1 = 14.496874999999998
x2 = 14.506874999999999
f(x1) = 5.253134765625003
f(x2) = 5.243172265625001
l = 1.008750000000001

5 итерация (a = 14.496874999999998, b = 15.00625):

c = 14.751562499999999
x1 = 14.746562499999998
x2 = 14.7565625
f(x1) = 5.064230566406251
f(x2) = 5.05926181640625
l = 0.5093750000000021

6 итерация (a = 14.746562499999998, b = 15.00625):

c = 14.876406249999999
x1 = 14.871406249999998
x2 = 

# Фибоначчи

In [19]:
print("Метод Фибоначчи, исходная функция (x - 15)^2 + 5")
a = 12
b = 20

epsilon = 0.01

func = lambda x: ((x - 15)**2) + 5

def fibonacci_list():
    res = [0, 1]
    count1 = 0
    count2 = 1
    fib = 1
    while not(fib >= ((b - a) / epsilon)):
        fib = res[count1] + res[count2]
        res.append(fib)
        count1 += 1
        count2 += 1
    return res

fibonacci_list = fibonacci_list()
print(fibonacci_list, len(fibonacci_list)) # [0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377, 610, 987, 1597] 18
l = b - a
count = 1

x1 = a + (fibonacci_list[len(fibonacci_list) - 2] / fibonacci_list[len(fibonacci_list) - 1]) * l
x2 = b - (fibonacci_list[len(fibonacci_list) - 2] / fibonacci_list[len(fibonacci_list) - 1]) * l
f1 = func(x1)
f2 = func(x2)
print("Итерация 0:")
print(f"a = {a}, b = {b}, l = {l}, x1 = {x1}, x2 = {x2}, f1 = {f1}, f2 = {f2}")

while (len(fibonacci_list) > 2):
    print(f"Итерация {count}:")
    if (f1 > f2):
        b = x1
        f1 = f2
        x1 = x2
        l = b - a
        x2 = a + (b - x1)
        f2 = func(x2)
        print(f"a = {a}, b = {b}, l = {l}, x1 = {x1}, x2 = {x2}, f1 = {f1}, f2 = {f2}")
    else:
        a = x2
        f2 = f1
        x2 = x1
        l = b - a
        x1 = b - (x2 - a)
        f1 = func(x1)
        print(f"a = {a}, b = {b}, l = {l}, x1 = {x1}, x2 = {x2}, f1 = {f1}, f2 = {f2}")
    del fibonacci_list[-1]
    count += 1
print(f"\nx* = {(a + b) / 2}\nРезультат f(x*) = {func((a + b) / 2)}")

Метод Фибоначчи, исходная функция (x - 15)^2 + 5
[0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377, 610, 987] 17
Итерация 0:
a = 12, b = 20, l = 8, x1 = 16.944275582573454, x2 = 15.055724417426546, f1 = 8.780207540991343, f2 = 5.003105210697528
Итерация 1:
a = 12, b = 16.944275582573454, l = 4.944275582573454, x1 = 15.055724417426546, x2 = 13.888551165146907, f1 = 5.003105210697528, f2 = 6.235318512496297
Итерация 2:
a = 13.888551165146907, b = 16.944275582573454, l = 3.0557244174265463, x1 = 15.777102330293815, x2 = 15.055724417426546, f1 = 5.6038880317480775, f2 = 5.003105210697528
Итерация 3:
a = 13.888551165146907, b = 15.777102330293815, l = 1.8885511651469074, x1 = 15.055724417426546, x2 = 14.609929078014176, f1 = 5.003105210697528, f2 = 5.152155324178871
Итерация 4:
a = 14.609929078014176, b = 15.777102330293815, l = 1.167173252279639, x1 = 15.331306990881444, x2 = 15.055724417426546, f1 = 5.109764322206917, f2 = 5.003105210697528
Итерация 5:
a = 14.609929078014176, b = 15

# Золотое сечение

In [20]:
def golden_ratio_iter(a, b, epsilon=0.01):
    t = 0.618
    L = b - a
    x1 = a + L * t
    x2 = b - L * t
    
    f1 = f(x1)
    f2 = f(x2)
    
    while L >= epsilon:
        if f1 > f2:
            b = x1
            f1 = f2
            x1 = x2
            L = b - a
            
            x2 = b - L * t
            f2 = f(x2)
        else:
            a = x2
            f2 = f1
            x2 = x1
            L = b - a
            
            x1 = a + L * t
            f1 = f(x1)
    
    return f((a + b) / 2)

a = 12
b = 20
print(golden_ratio_iter(a, b))

5.00000178203591


a = 12
b = 20
L = b - a = 8
t = 0.618

Шаг 0:
a = 12.000
b = 20.000
L = 8.000

x₁ = a + L·t = 12 + 8·0.618 = 12 + 4.944 = 16.944
x₂ = b - L·t = 20 - 8·0.618 = 20 - 4.944 = 15.056

f₁ = f(16.944) = (16.944-15)² + 5 = (1.944)² + 5 = 3.779 + 5 = 8.779
f₂ = f(15.056) = (15.056-15)² + 5 = (0.056)² + 5 = 0.003 + 5 = 5.003

Шаг1:
f₁ > f₂

a = a = 12.000
b = x₁ = 16.944
f₁ = f₂ = 5.003
x₁ = x₂ = 15.056
L = b - a = 16.944 - 12.000 = 4.944

x₂ = b - L·t = 16.944 - 4.944·0.618 = 16.944 - 3.055 = 13.889  
f₂ = f(13.889) = (13.889-15)² + 5 = (-1.111)² + 5 = 1.234 + 5 = 6.234

Шаг 2:
f₁ < f₂

a = x₂ = 13.889
b = b = 16.944
f₂ = f₁ = 5.003
x₂ = x₁ = 15.056
L = b - a = 16.944 - 13.889 = 3.055

x₁ = a + L·t = 13.889 + 3.055·0.618 = 13.889 + 1.888 = 15.777
f₁ = f(15.777) = (15.777-15)² + 5 = (0.777)² + 5 = 0.604 + 5= 5.604

O(log(t, (b-a)/epsilon))